### PR TITLE
archive flag for turning off GC

### PR DESCRIPTION
**Pull request by Kouprin**
Fixes https://github.com/nearprotocol/nearcore/issues/2094.

### CODE REVIEW

File: chain/client/src/types.rs [hunk 1]
```diff
@@ -136,6 +136,8 @@ pub struct ClientConfig {
     pub tracked_accounts: Vec<AccountId>,
     /// Shards that this client tracks
     pub tracked_shards: Vec<ShardId>,
+    /// Not clear old data, set `true` for achieve nodes.
```
Comment (ilblackdragon) [inline chain/client/src/types.rs]: Archive

Comment (Kouprin) [reply chain/client/src/types.rs]: oops ++

File: near/src/main.rs [hunk 2]
```diff
@@ -78,6 +78,7 @@ fn main() {
             .arg(Arg::with_name("network-addr").long("network-addr").help("Customize network listening address (useful for running multiple nodes on the same machine)").takes_value(true))
             .arg(Arg::with_name("rpc-addr").long("rpc-addr").help("Customize RPC listening address (useful for running multiple nodes on the same machine)").takes_value(true))
             .arg(Arg::with_name("telemetry-url").long("telemetry-url").help("Customize telemetry url").takes_value(true))
+            .arg(Arg::with_name("archive").long("archive").help("Keep old blocks in the storage (default false)").takes_value(true))
```
Comment (SkidanovAlex) [inline near/src/main.rs]: Why is `takes_value=true`, if it's a boolean?
Look at `fast` parameter above for the `init` on line 66.

Separately, no need to provide the default, it is common sense that if there's a flag, than omitting it is the reverse of its behavior (whomever added `produce_empty_blocks` above for some reason also chose to make it a string parameter, but booleans should be `takes_value=false`, and not specify default value in the docs).

**Comment by Kouprin** [comment]
I don't like the solution.
The problem is, if we _at least once_ wrongly run archive node without flag, we will lose all the data forever. People do mistakes, and I believe this case will happen one day.
@ilblackdragon @ailisp how can we store the flag permanently until it's dropped manually?

**Comment by Kouprin** [comment]
> I suggest we extend the idea of @Kouprin and @mfornet for the adversarial node:
> 
> We are going to have several types of nodes in our system:
> 
> * regular node (the only node for which we will be distributing the docker image);
> * archival node;
> * debug/diagnostics node;
> * adversarial node.
> 
> To run any of the above nodes, except the regular one, the user needs to provide the following three things:
> 
> * The code should be compiled with the feature flag that guards conditional compilation;
> * The node should be started with a special command line flag;
> * There should be an intent file next to the node.
> 
> For example, if we want to run adversarial node we should compile it with `adversarial_you_will_be_slashed` feature, then pass `--adversarial_you_will_be_slashed` flag and create `adversarial_you_will_be_slashed.txt` file before starting the node.
> 
> Pros:
> 
> * Uniform treatment of various types of the node:
>   
>   * Adversarial node and debug/diagnostics node use conditional compilation anyway;
> * Conditional compilation allows having optional external dependencies that might be used by heavy versions of the node, like "archival node" and "debug/diagnostics node";
> * Triple opt-in makes it really hard for someone to launch the wrong node accidentally. There is no need to have a clean-up step where we yank adversarial code when merging into master;
> 
> WDYT @Kouprin , @SkidanovAlex , @mfornet ?

Is it a separate issue, isn't it?

**Comment by Kouprin** [comment]
@nearmax I'm okay with accepting this as is now and creating a new issue for you idea.

**Comment by nearmax** [comment]
@Kouprin 
> Is it a separate issue, isn't it?

If we go with the above proposal then the code in this PR should be changed. Instead of threading `pub archive: bool,` flag everywhere we should use `#[cfg(feature = "archival_node")]` everywhere (which is btw better because threading requires increasing number of function arguments and struct fields).

**Comment by nearmax** [comment]
> @nearmax I'm okay with accepting this as is now and creating a new issue for you idea.

Ok

**Comment by Kouprin** [comment]
> @Kouprin
> 
> > Is it a separate issue, isn't it?
> 
> If we go with the above proposal then the code in this PR should be changed. Instead of threading `pub archive: bool,` flag everywhere we should use `#[cfg(feature = "archival_node")]` everywhere (which is btw better because threading requires increasing number of function arguments and struct fields).

There are different issues. This one is exactly for creating ability of system to turn off GC. Your suggestion is about how to make up the way of node running. This is great, but it's still different.

**Pull request merged**